keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,705 @@
+    Internal or External risks having some components misinterpret their
+    semantics.
+
+  - The vSphere provider claims that all IPs that it knows about are
+    *both* "Internal" and "External".
+
+  - In bare-metal environments, nodes have only a single IP, and that IP
+    is always called "Internal".
+
+[The documentation of node addresses]: https://kubernetes.io/docs/concepts/architecture/nodes/#addresses
+
+#### Backward Compatibility
+
+`Node.Status.Addresses` needs to stick around as long as `v1.Node`
+does, but it can be essentially frozen with its current semantics.
+Existing cloud providers may continue to set `Node.Status.Addresses`
+with values that are not entirely equivalent to the `Node.Status.IPs`
+values, for backward compatibility. For new cloud providers we may be
+able to have `Node.Status.Addresses` be automatically generated from
+`Node.Status.IPs` and `Node.Status.Hostname` by
+cloud-controller-manager.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+```
+<<[UNRESOLVED inconsistency ]>>
+Security issues related to Node.Status.Addresses and Node.Status.IPs
+containing different values?
+<<[/UNRESOLVED]>>
+```
+
+## Design Details
+
+### Implementation
+
+#### Hostname
+
+With the deprecation of `Node.Status.Addresses`, this value will be
+moved to `Node.Status.Hostname`, but does not otherwise need to be
+changed.
+
+```
+<<[UNRESOLVED hostname-label ]>>
+Is it a bug that `Node.Status.Hostname`/`NodeHostName` and the
+`"kubernetes.io/hostname"` label don't necessarily match? (NB: This
+is an existing behavior.)
+<<[/UNRESOLVED]>>
+```
+
+#### IPs
+
+The other information in `Node.Status.Addresses` will be moved to a
+new `Node.Status.IPs` array:
+
+    type NodeStatus struct {
+            ...
+
+            // List of the node's IPs. The first IP in the list is the "primary" node IP
+            // (eg, the hostIP value of pods on that node). If the second IP is of the
+            // opposite IP address family from the first (eg, IPv6 vs IPv4), then it is
+            // considered the primary IP of that address family. Other than that, the
+            // order of elements carries no meaning. Note that there may be a mix of IPv4
+            // and IPv6 addresses here regardless of whether the cluster supports both
+            // address families.
+            IPs []NodeIP `json:"ips"`
+
+            ...
+    }
+
+(Unlike with `Addresses`, this has no declared `patchStrategy` and thus
+defaults to `replace`, which is correct since it is solely maintained by
+a single component (kubelet or cloud-controller-manager), which always
+updates the entire field at once.)
```

Comment:
No. If the client and server don't agree as to what patchStrategies are in effect then the server will interpret the patch differently from how the client intended, so things would break terribly during an upgrade when the apiserver had the new semantics and kubelet still had the old ones.